plugins/repository-gcs/src/main/java/io/crate/gcs/GCSBlobStore.java

```diff
@@ -54,11 +104,283 @@ public BlobContainer blobContainer(BlobPath path) {
+                    throw new FileAlreadyExistsException(blobInfo.getBlobId().getName(), null, se.getMessage());
+                }
+                if (storageException != null) {
+                    se.addSuppressed(storageException);
+                }
+                throw se;
+            }
+        }
+        throw storageException;
+    }
+
+    /**
+     * Uploads a blob using the "multipart upload" method (a single
+     * 'multipart/related' request containing both data and metadata. The request is
+     * gziped), see:
+     * https://cloud.google.com/storage/docs/json_api/v1/how-tos/multipart-upload
+     *
+     * @param blobInfo            the info for the blob to be uploaded
+     * @param inputStream         the stream containing the blob data
+     * @param blobSize            the size
+     * @param failIfAlreadyExists whether to throw a FileAlreadyExistsException if the given blob already exists
+     */
+    private void writeBlobMultipart(BlobInfo blobInfo, InputStream inputStream, long blobSize, boolean failIfAlreadyExists)
+        throws IOException {
+        assert blobSize <= getLargeBlobThresholdInBytes() : "large blob uploads should use the resumable upload method";
+        byte[] buffer = new byte[Math.toIntExact(blobSize)];
+        inputStream.readNBytes(buffer, 0, buffer.length);
+        try {
+            final Storage.BlobTargetOption[] targetOptions = failIfAlreadyExists
+                ? new Storage.BlobTargetOption[]{Storage.BlobTargetOption.doesNotExist()}
+                : new Storage.BlobTargetOption[0];
+            client().create(blobInfo, buffer, targetOptions);
+        } catch (StorageException se) {
+            if (failIfAlreadyExists && se.getCode() == HTTP_PRECON_FAILED) {
+                throw new FileAlreadyExistsException(blobInfo.getBlobId().getName(), null, se.getMessage());
+            }
+            throw se;
         }
     }
+
+    void deleteDirectory(String pathStr) throws IOException {
+        Page<Blob> page = client().list(bucketName, BlobListOption.prefix(pathStr));
+        do {
+            Collection<String> blobsToDelete = new ArrayList<>();
+            AtomicLong blobsDeleted = new AtomicLong(0L);
+            AtomicLong bytesDeleted = new AtomicLong(0L);
+            page.getValues().forEach(b -> {
+                blobsToDelete.add(b.getName());
+                blobsDeleted.incrementAndGet();
+                bytesDeleted.addAndGet(b.getSize());
+            });
+            deleteBlobsIgnoringIfNotExists(blobsToDelete);
+            page = page.getNextPage();
+        } while (page != null);
+    }
+
+    void deleteBlobsIgnoringIfNotExists(Collection<String> blobNames) throws IOException {
+        if (blobNames.isEmpty()) {
+            return;
+        }
+        List<BlobId> blobIdsToDelete = blobNames.stream().map(blob -> BlobId.of(bucketName, blob)).collect(Collectors.toList());
```

Comment:
could use `toList()` instead (would remove the idea warning)
```suggestion
        List<BlobId> blobIdsToDelete = blobNames.stream().map(blob -> BlobId.of(bucketName, blob)).toList();
```